File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-11/28-Maximum-Sliding-Window.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        '''
        T: O(n log k) and S: O(k)
        
        result = []
        for i in range(len(nums) - k + 1):
            numsK = nums[i: i + k]
            heapq._heapify_max(numsK)
            result.append(numsK[0])
        return result
        
        T: O(n) and S: O(k)
        '''
        result = []
        q = deque()
        
        for i in range(len(nums)):            
            while q and (nums[q[-1]] < nums[i]):
                q.pop()
            if q and (i - q[0] >= k):
                q.popleft()
            q.append(i)
            result.append(nums[q[0]])
            
        return result[k-1:]
        

File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-11/28-Maximum-Sliding-Window.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        /**
        * T: O(n) and S: O(k)
        */
        List<Integer> result = new ArrayList<>();
        ArrayDeque<Integer> q = new ArrayDeque<>();
        
        for (int i = 0; i < nums.length; i++) {
            while (!q.isEmpty() && (nums[q.getLast()] < nums[i])) {
                q.removeLast();
            }
            if (!q.isEmpty() && (i - q.getFirst() >= k)) {
                q.removeFirst();
            }
            q.add(i);
            result.add(nums[q.getFirst()]);
        }
        int[] res = new int[result.size() - k + 1];
        for (int i = k - 1; i < result.size(); i++) {
            res[i-k+1] = result.get(i);
        }
        return res;
    }
}
